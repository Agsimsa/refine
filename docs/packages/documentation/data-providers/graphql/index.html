<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/docs/packages/data-providers/graphql/">
    <link rel="canonical" href="/docs/packages/data-providers/graphql/" />
  </head>
  <script>
    window.location.href = '/docs/packages/data-providers/graphql/' + window.location.search + window.location.hash;
  </script>
</html>