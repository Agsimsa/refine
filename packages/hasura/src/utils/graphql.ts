import { DocumentNode, visit } from "graphql";

export const isMutation = (documentNode: DocumentNode) => {
    let isMutation = false;

    visit(documentNode, {
        OperationDefinition: {
            enter(node) {
                if (node.operation === "mutation") {
                    isMutation = true;
                }
            },
        },
    });

    return isMutation;
};
