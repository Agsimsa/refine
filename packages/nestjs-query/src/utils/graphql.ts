import { FieldNode, DocumentNode, visit } from "graphql";

export const isNodesField = (node: FieldNode) => {
    return (
        node.selectionSet &&
        node.selectionSet.selections.length &&
        node.selectionSet.selections[0].kind === "Field" &&
        node.selectionSet.selections[0].name.value === "nodes"
    );
};

export function removeNodesField(inputString: string) {
    if (inputString.startsWith("nodes")) {
        const lines = inputString.split("\n");

        const filteredLines = lines
            .filter((line) => line.trim() !== "nodes {" && line.trim() !== "}")
            .map((line) => line.trim());

        return filteredLines.join("\n");
    }

    return inputString;
}

export function fieldsToString(documentNode: DocumentNode) {
    const fieldLines: string[] = [];
    let depth = 0;
    let isNestedField = false;

    visit(documentNode, {
        Field: {
            enter(node) {
                if (node.selectionSet && !isNestedField) {
                    isNestedField = true;
                    return;
                }

                if (isNodesField(node) || node.name.value === "totalCount")
                    return;

                if (node.selectionSet) {
                    depth++;
                } else {
                    fieldLines.push(`${"  ".repeat(depth)}${node.name.value}`);
                }
            },
            leave(node) {
                if (node.selectionSet && isNestedField) {
                    const nestedFieldLines = fieldLines.splice(
                        -node.selectionSet.selections.length,
                    );
                    if (depth > 0) {
                        fieldLines.push(
                            `${"  ".repeat(depth - 1)}${
                                node.name.value
                            } {\n${nestedFieldLines.join("\n")}\n${"  ".repeat(
                                depth - 1,
                            )}}`,
                        );
                    }
                    if (depth > 0) {
                        depth--;
                    }
                    if (depth === 0) {
                        isNestedField = false;
                    }
                }
            },
        },
    });

    if (fieldLines.length && fieldLines[0].startsWith("nodes")) {
        fieldLines[0] = removeNodesField(fieldLines[0]);
    }

    return fieldLines.join("\n").trim();
}

export const isMutation = (documentNode: DocumentNode) => {
    let isMutation = false;

    visit(documentNode, {
        OperationDefinition: {
            enter(node) {
                if (node.operation === "mutation") {
                    isMutation = true;
                }
            },
        },
    });

    return isMutation;
};
