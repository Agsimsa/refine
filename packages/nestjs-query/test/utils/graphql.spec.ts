import { FieldNode } from "graphql";
import gql from "graphql-tag";

import {
    fieldsToString,
    isMutation,
    isNodesField,
    removeNodesField,
} from "../../src/utils/graphql";

describe("isNodesField", () => {
    it('should return true for a node with a "nodes" field', () => {
        const query = gql`
            {
                users {
                    nodes {
                        id
                    }
                }
            }
        `;
        const node = (query.definitions[0] as any).selectionSet
            .selections[0] as FieldNode;
        expect(isNodesField(node)).toBeTruthy();
    });

    it('should return false for a node without a "nodes" field', () => {
        const query = gql`
            {
                users {
                    id
                }
            }
        `;
        const node = (query.definitions[0] as any).selectionSet
            .selections[0] as FieldNode;
        expect(isNodesField(node)).toBeFalsy();
    });
});

describe("removeNodesField", () => {
    it("should remove the nodes field from the string", () => {
        const inputString = `nodes {\nid\nname\navatarUrl\n}`;
        const expectedOutput = `id\nname\navatarUrl`;

        expect(removeNodesField(inputString)).toEqual(expectedOutput);
    });

    it("should return the same string if no nodes field is present", () => {
        const inputString = `users {\n  id\n}`;
        expect(removeNodesField(inputString)).toEqual(inputString);
    });
});

describe("fieldsToString", () => {
    it("UsersSelect", () => {
        const query = gql`
            query UsersSelect {
                users {
                    nodes {
                        id
                        name
                        avatarUrl
                    }
                }
            }
        `;
        const expectedOutput = `id\nname\navatarUrl`;
        expect(fieldsToString(query)).toEqual(expectedOutput);
    });

    it("ContactShow", () => {
        const query = gql`
            query ContactShow($id: ID!) {
                contact(id: $id) {
                    id
                    name
                    email
                    company {
                        id
                        name
                        avatarUrl
                    }
                    status
                    jobTitle
                    phone
                    timezone
                    avatarUrl
                    salesOwner {
                        id
                        name
                        avatarUrl
                    }
                    createdAt
                }
            }
        `;

        // TODO: Fix `salesOwner` absent from fields.
        // const expectedOutput = `id\nname\nemail\ncompany {\n  id\n  name\n  avatarUrl\n}\nstatus\njobTitle\nphone\ntimezone\navatarUrl\nsalesOwner {\n  id\n  name\n  avatarUrl\n}\ncreatedAt`;

        expect(fieldsToString(query)).toMatchSnapshot();
    });

    it("Upcoming Events", () => {
        const query = gql`
            query UpcomingEvents(
                $filter: EventFilter!
                $sorting: [EventSort!]
                $paging: OffsetPaging!
            ) {
                events(filter: $filter, sorting: $sorting, paging: $paging) {
                    nodes {
                        id
                        title
                        color
                        startDate
                        endDate
                    }
                    totalCount
                }
            }
        `;

        expect(fieldsToString(query)).toMatchSnapshot();
    });

    it("CompaniesTable", () => {
        const query = gql`
            query CompaniesTable(
                $filter: CompanyFilter!
                $sorting: [CompanySort!]!
                $paging: OffsetPaging!
            ) {
                companies(filter: $filter, sorting: $sorting, paging: $paging) {
                    nodes {
                        id
                        name
                        avatarUrl
                        dealsAggregate {
                            sum {
                                value
                            }
                        }
                        salesOwner {
                            id
                            name
                            avatarUrl
                        }
                        contacts {
                            nodes {
                                id
                                name
                                avatarUrl
                            }
                        }
                    }
                    totalCount
                }
            }
        `;

        expect(fieldsToString(query)).toMatchSnapshot();
    });

    it("CompanyTitleForm", () => {
        const query = gql`
            mutation CompanyTitleForm($input: UpdateOneCompanyInput!) {
                updateOneCompany(input: $input) {
                    id
                    name
                    avatarUrl
                    salesOwner {
                        id
                        name
                        avatarUrl
                    }
                }
            }
        `;

        expect(fieldsToString(query)).toMatchSnapshot();
    });

    it("CategoryEdit", () => {
        const query = gql`
            mutation CategoryEdit($input: UpdateOneCategoryInput!) {
                updateOneCategory(input: $input) {
                    id
                    title
                }
            }
        `;

        expect(fieldsToString(query)).toMatchSnapshot();
    });
});

describe("isMutation", () => {
    it("should return true for a mutation", () => {
        const mutation = gql`
            mutation {
                addUser {
                    id
                }
            }
        `;
        expect(isMutation(mutation)).toBeTruthy();
    });

    it("should return false for a query", () => {
        const query = gql`
            {
                users {
                    id
                }
            }
        `;
        expect(isMutation(query)).toBeFalsy();
    });
});
