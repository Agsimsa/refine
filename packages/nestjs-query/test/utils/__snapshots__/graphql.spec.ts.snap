// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldsToString CompaniesTable 1`] = `
"id
  name
  avatarUrl
  dealsAggregate {
    sum {
      value
    }
  }
  salesOwner {
    id
    name
    avatarUrl
  }
contacts {
  nodes {
    id
    name
    avatarUrl
  }
}"
`;

exports[`fieldsToString CompanyTitleForm 1`] = `
"id
name
avatarUrl
salesOwner {
  id
  name
  avatarUrl
}"
`;

exports[`fieldsToString ContactShow 1`] = `
"id
name
email
company {
  id
  name
  avatarUrl
}
status
jobTitle
phone
timezone
avatarUrl
createdAt"
`;

exports[`fieldsToString Upcoming Events 1`] = `
"id
title
color
startDate
endDate"
`;
